File: ios/AtlasCore/Sources/AtlasCore/Participant/Address.swift
import Foundation

/// Addressing + nameplate resolver — how a persona is FOUND, kept separate from how it's REACHED.
/// Swift parity with `backend/atlas/interop/address.py` (Python reference of record).
///
/// An address is `local@place`. `@place` says where to fetch a signed NAMEPLATE; `local` selects which.
/// A nameplate carries the persona key, a display name, a way-to-knock (receptive mode), and optionally
/// its self-published trust bundle — never a mailbox. Signed by the persona key, so the host is
/// untrusted. Findable (is there a nameplate) and receptive (can/how you may open a channel) are two
/// independent switches. Many addresses can point at one persona and converge into one inbox via a
/// PRIVATE owner-held routing table that senders never see.
public enum Address {

    public enum AddressError: Error { case malformed, verificationFailed }

    public enum ReceptiveMode: String, Sendable {
        case open = "open"
        case codeOnly = "code-only"
        case contactsOnly = "contacts-only"
        case closed = "closed"
    }

    public struct Parsed: Equatable {
        public let local: String
        public let place: String
        public var string: String { "\(local)@\(place)" }
    }

    public static func parse(_ s: String) throws -> Parsed {
        let t = s.trimmingCharacters(in: .whitespaces)
        let parts = t.components(separatedBy: "@")
        guard parts.count == 2, !parts[0].isEmpty, !parts[1].isEmpty else { throw AddressError.malformed }
        let place = parts[1].trimmingCharacters(in: CharacterSet(charactersIn: "/")).lowercased()
        guard !place.isEmpty else { throw AddressError.malformed }
        return Parsed(local: parts[0], place: place)
    }

    public static func nameplateURL(_ a: Parsed) -> String {
        "https://\(a.place)/.well-known/atlas/\(a.local).json"
    }

    public struct DiscoverySettings {
        public let findable: Bool
        public let receptive: ReceptiveMode
        public init(findable: Bool = false, receptive: ReceptiveMode = .closed) {
            self.findable = findable; self.receptive = receptive
        }
        public func acceptsContact(hasValidCode: Bool, isKnownContact: Bool) -> Bool {
            switch receptive {
            case .open: return true
            case .codeOnly: return hasValidCode
            case .contactsOnly: return isKnownContact
            case .closed: return false
            }
        }
    }

    private static func lp(_ b: Data) -> Data {
        var n = UInt32(b.count).bigEndian
        var out = Data(); withUnsafeBytes(of: &n) { out.append(contentsOf: $0) }; out.append(b)
        return out
    }

    public struct Nameplate {
        public let local: String
        public let place: String
        public let key: HybridSign.PublicKey
        public let displayName: String
        public let receptive: ReceptiveMode
        public let trustBundle: TrustPublish.TrustBundle?
        public var sig: Data = Data()

        public init(local: String, place: String, key: HybridSign.PublicKey, displayName: String,
                    receptive: ReceptiveMode, trustBundle: TrustPublish.TrustBundle? = nil, sig: Data = Data()) {
            self.local = local; self.place = place; self.key = key; self.displayName = displayName
            self.receptive = receptive; self.trustBundle = trustBundle; self.sig = sig
        }

        public func body() -> Data {
            let bundleTag = trustBundle?.body() ?? Data()
            return Primitives.H(Data("atlas/nameplate/v1".utf8), Address.lp(Data(local.utf8)),
                                Address.lp(Data(place.utf8)), Address.lp(key.encode()),
                                Address.lp(Data(displayName.utf8)), Address.lp(Data(receptive.rawValue.utf8)),
                                Address.lp(bundleTag))
        }

        public func toJSON() -> Data {
            var obj: [String: Any] = ["local": local, "place": place, "key": key.encode().hexString,
                                      "display_name": displayName, "receptive": receptive.rawValue,
                                      "sig": sig.hexString]
            if let b = trustBundle,
               let sub = try? JSONSerialization.jsonObject(with: b.toJSON()) {
                obj["trust_bundle"] = sub
            }
            return try! JSONSerialization.data(withJSONObject: obj, options: [.sortedKeys])
        }
    }

    public static func buildNameplate(_ persona: HybridSign.Keypair, address a: Parsed, displayName: String,
                                      settings: DiscoverySettings,
                                      trustBundle: TrustPublish.TrustBundle? = nil) throws -> Nameplate? {
        guard settings.findable else { return nil }
        var np = Nameplate(local: a.local, place: a.place, key: persona.publicKey, displayName: displayName,
                           receptive: settings.receptive, trustBundle: trustBundle)
        np.sig = try HybridSign.sign(persona, np.body())
        return np
    }

    public static func verifyNameplate(_ np: Nameplate) -> Bool {
        HybridSign.verify(np.key, np.body(), np.sig)
    }

    public static func parseNameplate(_ data: Data) throws -> Nameplate {
        guard let d = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let local = d["local"] as? String, let place = d["place"] as? String,
              let keyHex = d["key"] as? String, let key = HybridSign.PublicKey(encoded: Data(hex: keyHex)),
              let displayName = d["display_name"] as? String,
              let recRaw = d["receptive"] as? String, let receptive = ReceptiveMode(rawValue: recRaw),
              let sigHex = d["sig"] as? String else { throw AddressError.malformed }
        var bundle: TrustPublish.TrustBundle? = nil
        if let sub = d["trust_bundle"] {
            let subData = try JSONSerialization.data(withJSONObject: sub)
            bundle = try TrustPublish.parseTrustBundle(subData)
        }
        return Nameplate(local: local, place: place, key: key, displayName: displayName,
                         receptive: receptive, trustBundle: bundle, sig: Data(hex: sigHex))
    }

    public static func resolve(_ addressStr: String, fetch: (String) throws -> Data) throws -> Nameplate {
        let a = try parse(addressStr)
        let np = try parseNameplate(fetch(nameplateURL(a)))
        guard np.local == a.local, np.place == a.place, verifyNameplate(np) else {
            throw AddressError.verificationFailed
        }
        return np
    }

    /// The owner's PRIVATE map: which persona an address points at, and which real inbox a persona
    /// converges into. Never published — senders only see the persona face they were handed.
    public final class RoutingTable {
        private var addrToPersona: [String: Data] = [:]
        private var personaToInbox: [Data: Data] = [:]
        public init() {}

        public func point(_ addressStr: String, persona: HybridSign.PublicKey) {
            addrToPersona[addressStr] = persona.encode()
        }
        public func converge(_ persona: HybridSign.PublicKey, inbox: Data) {
            personaToInbox[persona.encode()] = inbox
        }
        public func personaFor(_ addressStr: String) -> Data? { addrToPersona[addressStr] }
        public func inboxFor(_ persona: HybridSign.PublicKey) -> Data? { personaToInbox[persona.encode()] }
        public func inboxForAddress(_ addressStr: String) -> Data? {
            guard let pk = addrToPersona[addressStr] else { return nil }
            return personaToInbox[pk]
        }
    }
}

File: backend/atlas/interop/address.py
"""Addressing + nameplate resolver — how a persona is FOUND, kept separate from how it's REACHED.

An address is `local@place` (ada@example.town, ada@example.id). The `@place` half says where to
fetch a signed NAMEPLATE; the `local` part selects which one. A nameplate carries the persona's key, a
display name, a way-to-knock (its receptive mode), and optionally its self-published trust bundle. It
does NOT carry a mailbox — discovery is stable and public; delivery rotates and stays private (see
net/privacy/mailbox and the rotating connect code). Because the nameplate is signed by the persona key,
the host is untrusted: self-host it, mirror it, or serve it from your own node, it verifies the same.

Two independent switches per persona:
  * FINDABLE — is there a published nameplate at all / are you discoverable. Off = no nameplate.
  * RECEPTIVE — can someone open a channel, and on what terms: OPEN / CODE_ONLY / CONTACTS_ONLY / CLOSED.
Findable does not imply receptive: you can be loudly discoverable with a closed door, or unlisted yet
reachable by anyone holding your code.

Many addresses can point at one persona, and one persona's mail can converge into your real inbox — but
that mapping is a PRIVATE, owner-held RoutingTable, never published, so senders learn only the persona
face they were given. Swift parity in Address.swift.
"""
from __future__ import annotations

import json
from dataclasses import dataclass
from enum import Enum
from typing import Callable, Dict, Optional

from ..crypto.primitives import H
from ..crypto.sign import HybridSigKeypair, HybridSigPublic, sign, verify
from .trust_publish import TrustBundle, parse_trust_bundle


class AddressError(Exception):
    ...


class ReceptiveMode(Enum):
    OPEN = "open"                    # anyone may open a channel
    CODE_ONLY = "code-only"          # only with a valid connect code (see the rotating connect code)
    CONTACTS_ONLY = "contacts-only"  # only already-accepted contacts
    CLOSED = "closed"                # not reachable; discovery only


@dataclass(frozen=True)
class Address:
    local: str
    place: str

    def __str__(self) -> str:
        return f"{self.local}@{self.place}"


def parse_address(s: str) -> Address:
    s = s.strip()
    if s.count("@") != 1:
        raise AddressError("an address is exactly local@place")
    local, place = s.split("@")
    if not local or not place:
        raise AddressError("both local and place are required")
    return Address(local=local, place=place.strip("/").lower())


def nameplate_url(addr: Address) -> str:
    """Where the place serves this local's nameplate — one host serves many locals, like email."""
    return f"https://{addr.place}/.well-known/atlas/{addr.local}.json"


@dataclass
class DiscoverySettings:
    """A persona's two independent switches."""
    findable: bool = False
    receptive: ReceptiveMode = ReceptiveMode.CLOSED

    def accepts_contact(self, *, has_valid_code: bool, is_known_contact: bool) -> bool:
        if self.receptive is ReceptiveMode.OPEN:
            return True
        if self.receptive is ReceptiveMode.CODE_ONLY:
            return has_valid_code
        if self.receptive is ReceptiveMode.CONTACTS_ONLY:
            return is_known_contact
        return False


@dataclass
class Nameplate:
    """The signed discovery record served at nameplate_url. Signed by the persona key; the host is
    untrusted. Carries the receptive mode (a way to knock), never a mailbox."""
    local: str
    place: str
    key: HybridSigPublic
    display_name: str
    receptive: ReceptiveMode
    trust_bundle: Optional[TrustBundle] = None
    sig: bytes = b""

    def body(self) -> bytes:
        def lp(b: bytes) -> bytes:
            return len(b).to_bytes(4, "big") + b
        bundle_tag = self.trust_bundle.body() if self.trust_bundle is not None else b""
        return H(b"atlas/nameplate/v1", lp(self.local.encode()), lp(self.place.encode()),
                 lp(self.key.encode()), lp(self.display_name.encode()),
                 lp(self.receptive.value.encode()), lp(bundle_tag))

    def to_json(self) -> bytes:
        obj = {"local": self.local, "place": self.place, "key": self.key.encode().hex(),
               "display_name": self.display_name, "receptive": self.receptive.value,
               "sig": self.sig.hex()}
        if self.trust_bundle is not None:
            obj["trust_bundle"] = json.loads(self.trust_bundle.to_json().decode())
        return json.dumps(obj, sort_keys=True, separators=(",", ":")).encode()


def build_nameplate(persona: HybridSigKeypair, *, address: Address, display_name: str,
                    settings: DiscoverySettings, trust_bundle: Optional[TrustBundle] = None
                    ) -> Optional[Nameplate]:
    """Publish a nameplate for a FINDABLE persona (None if it isn't findable — nothing to serve)."""
    if not settings.findable:
        return None
    np = Nameplate(local=address.local, place=address.place, key=persona.public,
                   display_name=display_name, receptive=settings.receptive, trust_bundle=trust_bundle)
    np.sig = sign(persona, np.body())
    return np


def verify_nameplate(np: Nameplate) -> bool:
    """The nameplate is signed by the very key it advertises (so the place can't forge it)."""
    return verify(np.key, np.body(), np.sig)


def parse_nameplate(data: bytes) -> Nameplate:
    d = json.loads(data)
    bundle = None
    if "trust_bundle" in d:
        bundle = parse_trust_bundle(json.dumps(d["trust_bundle"]).encode())
    return Nameplate(local=d["local"], place=d["place"],
                     key=HybridSigPublic.decode(bytes.fromhex(d["key"])),
                     display_name=d["display_name"], receptive=ReceptiveMode(d["receptive"]),
                     trust_bundle=bundle, sig=bytes.fromhex(d["sig"]))


def resolve(address_str: str, *, fetch: Callable[[str], bytes]) -> Nameplate:
    """Look up an address: fetch its nameplate, verify the signature, return it. `fetch` is injected
    (real HTTP in the app/node, a stub in tests). Raises AddressError if the signature doesn't check."""
    addr = parse_address(address_str)
    np = parse_nameplate(fetch(nameplate_url(addr)))
    if np.local != addr.local or np.place != addr.place or not verify_nameplate(np):
        raise AddressError("nameplate failed verification or does not match the address")
    return np


class RoutingTable:
    """The owner's PRIVATE map: which persona an address points at, and which real inbox a persona
    converges into. Never published — senders only ever see the persona face they were handed, so
    convergence into one inbox stays invisible to them."""

    def __init__(self) -> None:
        self._addr_to_persona: Dict[str, bytes] = {}      # address string -> persona key enc
        self._persona_to_inbox: Dict[bytes, bytes] = {}   # persona key enc -> private inbox id

    def point(self, address_str: str, persona: HybridSigPublic) -> None:
        self._addr_to_persona[address_str] = persona.encode()

    def converge(self, persona: HybridSigPublic, inbox_id: bytes) -> None:
        self._persona_to_inbox[persona.encode()] = inbox_id

    def persona_for(self, address_str: str) -> Optional[bytes]:
        return self._addr_to_persona.get(address_str)

    def inbox_for(self, persona: HybridSigPublic) -> Optional[bytes]:
        return self._persona_to_inbox.get(persona.encode())

    def inbox_for_address(self, address_str: str) -> Optional[bytes]:
        """The full private hop a sender never sees: address -> persona -> your real inbox."""
        pk = self._addr_to_persona.get(address_str)
        return self._persona_to_inbox.get(pk) if pk is not None else None
